tetragon: handle process threads in clone and process_{exec|exit} events

tixxdz opened the pull request:
This PR adds the TID the thread ID to Process object and then it adds support to clone and process_exec|exit events.

The TID makes more sense in kprobe and tracepoints, however since we share the Process object with process_exec, let's reuse it here and ensure that the TID is not zero during process_{exec|exit}

For more details, see commits.

kkourt commented on bpf/lib/process.h:
```diff
@@ -177,6 +177,7 @@ struct msg_clone_event {
 	struct msg_common common;
 	struct msg_execve_key parent;
 	__u32 pid;
+	__u32 tid;
```
I think here pid is always tid, right?
If so, let's just keep `pid`. There is no need to add a new field if it always contains the same information with an existing one.

kkourt replied:
OK, let's go with that.

tixxdz replied:
Thank you much appreciated for your time! so I made all the other requested changes

kkourt commented on pkg/process/process.go:
```diff
@@ -257,6 +257,9 @@ func AddCloneEvent(event *tetragonAPI.MsgCloneEvent) error {
 		pi.process.Pid = &wrapperspb.UInt32Value{Value: event.PID}
+		// Add TID of the thread leader so it will be cached. Then on non execve events
+		// the TID of the triggered events will be fetched on the fly since we do not track it.
+		pi.process.Tid = &wrapperspb.UInt32Value{Value: event.TID}
```
Will we use the cached thread id only for exec events? What happens if exec is not executed by the thread group leader?

That is process does:
```
ppid: fork()   -> pid1 is created (pid1 == tid1)
pid1: pthread_create -> spawns tid2
tid2: exec()
```

Will we get tid1 in the exec event then?

Since the tid is per-thread, while the cache we maintain is per process, it does not seem right to add the tid to the cache. I think we should always fill it from information we get from the bpf side.

kkourt replied:
I believe subsequent  commit "bpf:exec: send the TID per exec event" https://github.com/cilium/tetragon/pull/941/commits/947fde34923bafb001d2cbf7ab1276a308a7b7e3 deals with above case.

Let's not add the tid to the process cache then?

tixxdz replied:
> Will we use the cached thread id only for exec events? What happens if exec is not executed by the thread group leader?

Yes since the cached Process must have the TID anyway as it is [API](https://github.com/cilium/tetragon/blob/main/pkg/process/process.go#L42) we can't ignore it.
 
> That is process does:
> 
> ```
> ppid: fork()   -> pid1 is created (pid1 == tid1)
> pid1: pthread_create -> spawns tid2
> tid2: exec()
> ```
> 
> Will we get tid1 in the exec event then?
> 
> Since the tid is per-thread, while the cache we maintain is per process, it does not seem right to add the tid to the cache. I think we should always fill it from information we get from the bpf side.

Good point @kkourt , and as you point out it is handled by the "bpf:exec: send the TID per exec event" https://github.com/cilium/tetragon/commit/947fde34923bafb001d2cbf7ab1276a308a7b7e3 , the exec event will have the correct TID that is same of TGID, cause our hook is late.

During the execve hook the tid tgid transfert has already happened between the old thread group leader and the new thread leader that triggered the execve https://elixir.bootlin.com/linux/v6.0/source/fs/exec.c#L1128  , in the end from our perspective the tid tgid should be the same, from the kernel perspective a task IDs exchange happened. Will add a test to the exec to assert on that

tixxdz replied:
> I believe subsequent commit "bpf:exec: send the TID per exec event" [947fde3](https://github.com/cilium/tetragon/commit/947fde34923bafb001d2cbf7ab1276a308a7b7e3) deals with above case.
> 
> Let's not add the tid to the process cache then?

You mean part of the clone event keep it nil? what's the point of that?

The process key added to the cache is the execID that is `event.PID and event.Ktime`  , the TID has no inference there, am I missing something?

tixxdz replied:
> Let's not add the tid to the process cache then?

Got your point now @kkourt ;-) yes ! but I don't think it will help us to improve things, beside see this comment https://github.com/cilium/tetragon/pull/941#issuecomment-1523297606 for the next PR kprobes and tracepoints and for out of order events, reading back the process from the cache to construct kprobe and tracepoint will return a process.TID == 0 and we have to overwrite it anyway with the one recorded from BPF during kprobe...

So I'm thinking what's the value of not pushing it, let me see try another version maybe you have a point and it is easy to read that way?

tixxdz replied:
Ok updated per your request to not add it into the cache, after more thinking I kept only from BPF side and will assert on tests that we do send the correct information but also added debug message in user space to ensure we log in case a mismatch.

kkourt commented on pkg/grpc/exec/exec.go:
```diff
@@ -273,14 +273,15 @@ func (msg *MsgCloneEventUnix) Cast(o interface{}) notify.Message {
 			Pid:       &wrapperspb.UInt32Value{Value: event.ProcessKey.Pid},
 			StartTime: ktime.ToProto(event.ProcessKey.Ktime),
 		}
+		process.UpdateEventProcessTid(tetragonProcess, event.Info.Tid)
```
Shouldn't we call UpdateEventProcessTid for other paths as well?

tixxdz replied:
For the `GetProcessExit->GetParentProcessInternal()` it should return a process that has the TID set from the exec event or from the clone one. Otherwise if we miss the cache we use the one that BPF recorded. These are the only path here

tixxdz replied:
Ok @kkourt , updated and made it simpler that all paths use the `UpdateEventProcessTid()`

kkourt commented on contrib/tester-progs/threads-tester.c:
```diff
@@ -0,0 +1,89 @@
+#define _GNU_SOURCE
+
+#define _GNU_SOURCE
```
nit: This seems to be defined twice.

tixxdz replied:
thanks, removed

kkourt commented on pkg/testutils/threadevents.go:
```diff
@@ -0,0 +1,63 @@
+	Thread1Re      = regexp.MustCompile(`Thread 1:\t\(pid:(\d+), tid:(\d+), ppid:(\d+)\)\t`)
+)
+
+func (tti *ThreadTesterInfo) ParseLine(l string) error {
```
Is there a reason for adding this to testutils? Are there any other users than threads_test?

tixxdz replied:
Ah yes, in the kprobe test of the follow up PR  https://github.com/cilium/tetragon/pull/946/commits/4bee439cee92c5447055a1c1390e02013e8875b7  , guess that's fine

tpapagian commented on pkg/process/process.go:
```diff
@@ -257,6 +257,9 @@ func AddCloneEvent(event *tetragonAPI.MsgCloneEvent) error {
 		pi.process.Pid = &wrapperspb.UInt32Value{Value: event.PID}
+		// Add TID of the thread leader so it will be cached. Then on non execve events
+		// the TID of the triggered events will be fetched on the fly since we do not track it.
+		pi.process.Tid = &wrapperspb.UInt32Value{Value: event.TID}
```
I don't have a strong opinion on adding `tid` into the process cache or not. We have to override that whenever we are going to generate any user events. Based on that we don't need to set this here, but it also does not cause any issues. 

Maybe for debugging purposes, we can set this here to an invalid value (i.e. `UINT32_MAX`?) and we should also check that we don't generate any events with that (wrong) value.

tixxdz replied:
> I don't have a strong opinion on adding `tid` into the process cache or not. We have to override that whenever we are going to generate any user events. Based on that we don't need to set this here, but it also does not cause any issues.
> 
> Maybe for debugging purposes, we can set this here to an invalid value (i.e. `UINT32_MAX`?) and we should also check that we don't generate any events with that (wrong) value.

That makes sense indeed, @kkourt had same suggestion to not put into the cache but have that value 0 , then later we can use it to know that we didn't add the execve event as an indication! so will use 0 as it is an invalid pid, and see how to add a final log message at the execve to catch/assert on this eventually!

For the clone event sending TID, we had a long discussion and I don't think it hurts having it, especially since I added BPF tests and debug messages to assert on it, and the updated version is clear about what we expect. So I guess I will keep it! thank you @tpapagian !

tpapagian reviewed (commented):
Thanks for that PR! This will be really useful especially in the case of `kprobe`/`tracepoint` events as you do in https://github.com/cilium/tetragon/pull/946.

I believe that it will work as expected but I have some comments on making that a bit more clear to argue about that.

tixxdz commented on vendor/github.com/cilium/tetragon/api/v1/tetragon/tetragon.proto:
```diff
@@ -95,6 +95,7 @@ message Process {
     uint32 refcnt = 13;
     Capabilities cap = 14;
     Namespaces ns = 15;
+    google.protobuf.UInt32Value tid = 16;
```
Indeed, done!

tixxdz commented on bpf/process/bpf_fork.c:
```diff
@@ -72,6 +75,10 @@ BPF_KPROBE(event_wake_up_new_task, struct task_struct *task)
 			.pid = curr->key.pid,
+			/* Explicitly use the TGID for TID as we generate one
+			 * event per process thread group.
+			 */
```
Thank you! added the comment

tixxdz commented on bpf/process/bpf_execve_event.c:
```diff
@@ -124,8 +124,21 @@ event_filename_builder(void *ctx, struct msg_process *curr, __u32 curr_pid, __u3
+	 * consistent with other BPF events.
+	 */
 	curr->pid = curr_pid;
+	curr->tid = curr_pid;
```
Indeed, bpf code changed and I'm able to load it successfully, I had two bpf errors during this series with kprobes.

So, removed that part, used one argument and shifted the related bytes accordingly. Thank you!

tixxdz commented on pkg/grpc/exec/exec.go:
```diff
@@ -273,14 +273,15 @@ func (msg *MsgCloneEventUnix) Cast(o interface{}) notify.Message {
 			Pid:       &wrapperspb.UInt32Value{Value: event.ProcessKey.Pid},
 			StartTime: ktime.ToProto(event.ProcessKey.Ktime),
 		}
+		process.UpdateEventProcessTid(tetragonProcess, event.Info.Tid)
```
Thank you @tpapagian alright updated it

tixxdz commented:
> > Thanks!
> > The patch looks good to me, overall. Having this will certainly make our events more robust.
> > I have some nits, but the main thing I'd like to ensure we do right is _always_ using the thread id from the bpf side rather than the one from the cache. This is something you mention in the comments, and I fully agree but it's not 100% clear to me that it is reflected in the implementation.
> 
> Sure that's the aim! so for this PR that handles process_exec/exit we do use the one from BPF side, let's break it please:
> 
>     * process_exec is received:
>       `(msg *MsgExecveEventUnix) HandleMessage()->GetProcessExec(msg, true)->process.AddExecEvent()->GetProcess(process MsgProcess)-> Tid: &wrapperspb.UInt32Value{Value: process.TID},` , so  AddExecEvent() constructs it, add it to the cache and return it for the event , we should be good.
> 
>     * process_exit with a cache hit, then good we just read it from the cache and that is same value that was added by Clone or Execve one. If an execve happens by another thread other than the thread leader, then tgid and tid transfert should happen and the thread exec'in is the new thread leader with same old values , but we still record that from BPF and return it to update the cache entry.
> 
>     * process_exit without a cache hit: then we just use the TID returned from BPF side.
> 
> 
> > Specifically, we seem to be adding the thread id in the process cache. Is there a reason for it? Shouldn't we always use the thread id we get from the bpf side?
> 

So I updated the series and converted the process_exit to always use the bpf TID for both cache hit or miss paths.

Now all events use BPF value, the execve is same one, just the implementation happens to be like that, plus I added warning to ensure that we log in case TID is not set and events are being reported without it..

Much appreciated for your time @kkourt @tpapagian ;-)

tpapagian reviewed (approved):
Thanks for the updates based on the comments. I think now this PR LGTM.

kkourt reviewed (approved):
🚀
